lib/jsdom/browser/location.js

```diff
@@ -1,103 +1,67 @@
+  this[document] = relevantDocument;
+  whatwgURL.setTheInput(this, urlString);
 
-    this._url = nodeURL.parse(resolveHref(oldUrl, val));
-    var newUrl = this._url.href;
-    var newProtocol = this._url.protocol;
-    var newHost = this._url.host;
-    var newPathname = this._url.pathname;
-    var newHash = this._url.hash || "";
+  try {
+    this[oldParsedURL] = new URL(urlString);
+  } catch (e) {
+    this[oldParsedURL] = {};
+  }
+}
 
-    if (oldProtocol === newProtocol && oldHost === newHost && oldPathname === newPathname && oldHash !== newHash) {
-      this._signalHashChange(oldUrl, newUrl);
-    } else {
-      notImplemented("location.href reloading", this._document._defaultView);
-    }
+whatwgURL.mixinURLUtils(
+  Location.prototype,
+  function getTheBase() {
+    return getDocumentBaseUrl(this[document]);
   },
+  function updateSteps() {
+    if (this[oldParsedURL].protocol !== this.protocol ||
+        this[oldParsedURL].username !== this.username ||
+        this[oldParsedURL].password !== this.password ||
+        this[oldParsedURL].hostname !== this.hostname ||
+        this[oldParsedURL].port !== this.port ||
+        this[oldParsedURL].pathname !== this.pathname ||
+        this[oldParsedURL].search !== this.search) {
+      notImplemented("navigation via the location interface", this[document]._defaultView);
+    } else if (this[oldParsedURL].hash !== this.hash) {
+      const window = this[document].defaultView;
+      const ev = new window.HashChangeEvent("hashchange", {
+        bubbles: true,
+        cancelable: false,
+        oldURL: this[oldParsedURL].href,
+        newURL: this.href
+      });
 
-  set hash(val) {
-    var oldUrl = this._url.href;
-    var oldHash = this._url.hash || "";
-
-    if (val.lastIndexOf("#", 0) !== 0) {
-      val = "#" + val;
+      setTimeout(function () {
+        window.dispatchEvent(ev);
+      }, 0);
     }
 
-    this._url = nodeURL.parse(resolveHref(oldUrl, val));
-    var newUrl = this._url.href;
-    var newHash = this._url.hash || "";
-
-    if (oldHash !== newHash) {
-      this._signalHashChange(oldUrl, newUrl);
-    }
-  },
-
-  set search(val) {
-    var oldUrl = this._url.href;
-    var oldHash = this._url.hash || "";
-    if (val.length) {
-      if (val.lastIndexOf("?", 0) !== 0) {
-        val = "?" + val;
-      }
-      this._url = nodeURL.parse(resolveHref(oldUrl, val + oldHash));
-    } else {
-      this._url = nodeURL.parse(oldUrl.replace(/\?([^#]+)/, ""));
-    }
-  },
+    this[oldParsedURL] = new URL(this.href);
+    this[document]._URL = this.href;
+  }
+);
 
-  replace: function (val) {
-    this.href = val;
-  },
+Location.prototype.assign = function () {
+  notImplemented("location.assign", this[document]._defaultView);
+};
 
-  toString: function () {
-    return this._url.href;
-  },
+Location.prototype.replace = function (url) {
+  // This is nowhere near spec compliant, but has worked so far.
+  whatwgURL.setTheInput(this, url);
```

Comment:
Not quite for replace. Replace generally does not cause a navigation. (It does, but with the replacement flag set.) The current code implements it as `this.href = val`.